LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs

```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
             return false;
         }
 
+        readonly SemaphoreSlim receiveAsyncLock = new SemaphoreSlim(1, 1);
+        TaskCompletionSource<Message> receiveAsyncTaskCompletionSource;
+        int receiveAsyncTaskWaitCount = 0;
+
         public async Task<Message> ReceiveAsync(TimeSpan timeout)
+        {
+            var isUsingPendingRequest = true;
+
+            TaskCompletionSource<Message> localPendingReceiveAsync = null;
+
+            await this.receiveAsyncLock.WaitAsync();
+
+            try
+            {
+                if ((localPendingReceiveAsync = this.receiveAsyncTaskCompletionSource) == null)
+                {
+                    localPendingReceiveAsync = this.SetupNewReceiveAsyncTaskCompletionSource(timeout);
+                    isUsingPendingRequest = false;
+                }
+            }
+            finally
+            {
+                this.receiveAsyncTaskWaitCount++;
+                this.receiveAsyncLock.Release();
+            }
+
+            if (isUsingPendingRequest)
+            {
+                Logger.Log(this.devEUI, $"checking cloud to device message for {timeout}, reusing pending request", LogLevel.Debug);
+            }
+
+            using (var cts = new CancellationTokenSource())
+            {
+                var timer = Task.Delay(timeout, cts.Token);
+                var winner = await Task.WhenAny(localPendingReceiveAsync.Task, timer);
+                if (winner == localPendingReceiveAsync.Task)
+                {
+                    // Cancel the timer tasks so that it does not fire
+                    cts.Cancel();
+
+                    Task<Message> singleFinished;
+                    await this.receiveAsyncLock.WaitAsync();
+                    try
+                    {
+                        if (localPendingReceiveAsync == this.receiveAsyncTaskCompletionSource)
+                        {
+                            Logger.Log(this.devEUI, $"task ReceiveAsync returned before timeout", LogLevel.Debug);
```

Comment:
Do you want to keep that?

Reply:
I will leave verbose log until release 1.0.3